reconcile/utils/oc.py

```diff
@@ -178,30 +241,46 @@ def remove_last_applied_configuration(self, namespace, kind, name):
                'kubectl.kubernetes.io/last-applied-configuration-']
         self._run(cmd)
 
-    @elapsed_seconds_from_commit_metric
+    def _msg_to_process_reconcile_time(self, namespace, resource):
+        return OCDecoratorMsg(
+            namespace=namespace,
+            resource=resource,
+            server=self.server,
+            slow_oc_reconcile_threshold=self.slow_oc_reconcile_threshold,
+            is_log_slow_oc_reconcile=self.is_log_slow_oc_reconcile)
+
+    @OCDecorators.process_reconcile_time
     def apply(self, namespace, resource):
-        self._log_slow_oc_reconcile(namespace, resource)
+        """ Runs oc apply. Returns nothing """
         cmd = ['apply', '-n', namespace, '-f', '-']
-        self._run(cmd, stdin=resource, apply=True)
+        self._run(cmd, stdin=resource.toJSON(), apply=True)
+        # This return will be removed by the last decorator
+        return self._msg_to_process_reconcile_time(namespace, resource.body)
 
-    @elapsed_seconds_from_commit_metric
+    @OCDecorators.process_reconcile_time
     def create(self, namespace, resource):
-        self._log_slow_oc_reconcile(namespace, resource)
+        """ Runs oc create. Returns nothing """
         cmd = ['create', '-n', namespace, '-f', '-']
-        self._run(cmd, stdin=resource, apply=True)
+        self._run(cmd, stdin=resource.toJSON(), apply=True)
+        # This return will be removed by the last decorator
+        return self._msg_to_process_reconcile_time(namespace, resource.body)
 
-    @elapsed_seconds_from_commit_metric
+    @OCDecorators.process_reconcile_time
     def replace(self, namespace, resource):
-        self._log_slow_oc_reconcile(namespace, resource)
+        """ Runs oc replace. Returns nothing """
         cmd = ['replace', '-n', namespace, '-f', '-']
-        self._run(cmd, stdin=resource, apply=True)
+        self._run(cmd, stdin=resource.toJSON(), apply=True)
+        # This return will be removed by the last decorator
+        return self._msg_to_process_reconcile_time(namespace, resource.body)
```

Comment:
whant heppens if a function returns something that should go through + the message to the decorator? example:
```python
    return actual_return, OCDecoratorMsg(...)
```
I guess what I'm trying to say is: the decorator should be totally transparent. even the message to the decorator should go through it and be present on the other side, but I'd be fine with not having the message going through. In any case, actual returns have to.

Reply:
the thing is that no modifying function in this class actually returns anything. I was trying to preserve that behaviour as much as possible. It looked pretty weird to me just return the decorator message but I must admit this is my first time dealing with decorators in Python so I lack the sense of best practices.

What I can do is making the decorator check for the last element of the return value and if it is an instance of the `OCDecoratorMsg` do stuff. I will return the contents of the call unmodified as in this case we don't really care, no one checks it.

I guess what you want as decorator is a function that can be consumed everywhere even by mistake whereas I'm thinking about a function that I know where it is called from. It doesn't hurt to be extra defensive, I guess :)

Reply:
What I'm thinking is that we can, at some point, intentionally use the decorator t odecorate a function that returns something. And, even then, both the decorator and the function should keep working as expected.

Reply:
I buy your argument ;-)

Reply:
To be very accurate, even these functions you're decorating, they return something. They return `None`. Your decorator also returns `None` when there's a message to the decorator, so this case is covered.

My concern is only when we try to decorate a function that has already an explicit return AND we want to track its reconcile time, meaning we have to add the message to the decorator in the existing explicit return.

Does that make sense?

Reply:
cool cool :)

Reply:
Fixed in https://github.com/app-sre/qontract-reconcile/pull/1382/commits/8204c5c61390d5fe750c2f576631eab860d06b64. This is the more generic I can imagine this 😂

Reply:
It's nice. And, most importantly, it works :)